File: modules/create.go
package modules

import (
	"github.com/sendoushi/bedrock-utils/modules/tools"
	"path"
)

// ---------------------------------
// Vars

// CreateStruct struct for the project
type CreateStruct struct {
	Dest  string `toml:"destination"`
	Type  string
	Order int
	Env   string
	Sys   string
}

// ---------------------------------
// Public functions

// CreateTask for init
func CreateTask(config []CreateStruct, commandType string, order int, env string, sys string) {
	if commandType != "init" {
		return
	}

	for _, task := range config {
		var shouldContinue bool
		if task.Order, task.Env, task.Sys, shouldContinue = tools.InitDecision(
			task.Order, task.Env, task.Sys, order, env, sys,
		); shouldContinue {
			continue
		}

		tools.Log("create", task.Type)
		_, err := CreateProject(task)

		// Log whatever
		tools.LogErr("create", err)
	}
}

// CreateProject creates a project
func CreateProject(module CreateStruct) (log string, err error) {
	modulesFiles, _ := tools.GetGlob(path.Join(CmdDir, "external/create", module.Type, "*"))
	files := []FileStruct{}

	// Go through each in the glob
	for _, file := range modulesFiles {
		files = append(files, FileStruct{Src: file, Dest: module.Dest, Force: true})
	}

	// Now lets run!
	FileTask(files, "copy", 0, "", "")

	return log, err
}

// ---------------------------------
// Private functions

File: modules/create.js
'use strict';

// -----------------------------------------
// IMPORTS

var path = require('path');
var Joi = require('joi');
var Promise = require('bluebird');

var tools = require('./tools/main');
var validate = tools.validate;
var file = require('./file');

// -----------------------------------------
// VARS

var struct = Joi.object().keys({
    dest: Joi.string().required(), // `toml:"destination"`
    type: Joi.string().required(),
    order: Joi.number().default(0),
    env: Joi.string().allow('').default(''),
    cmd: Joi.string().allow('').default(''),
    sys: Joi.string().default('all')
});

// -----------------------------------------
// PRIVATE FUNCTIONS

// -----------------------------------------
// PUBLIC FUNCTIONS

/**
 * Creates a project
 * @param  {object} module
 */
function project(module) {
    var modules;

    validate.type({ module: module }, { module: struct });

    // Now lets run!
    modules = tools.getGlob(path.join(__dirname, 'external/create', module.type, '**/*'));
    modules.forEach(function (fileObj) {
        var dest = path.join(module.dest, fileObj.relative);

        // Lets copy
        file.copy({ src: fileObj.absolute, dest: tools.getAbsolute(dest) });
    });
}

/**
 * Create task for init
 * @param  {object} bedrockObj
 * @param  {object} config
 */
function task(bedrockObj, config) {
    var promises = [];

    tools.setModule('create');
    validate.type(
        { config: config },
        { config: Joi.array().items(struct) }
    );

    // Go through each task
    config.forEach(function (configTask) {
        var shouldContinue = tools.decide(bedrockObj, configTask);
        if (shouldContinue) {
            return;
        }

        tools.log(configTask.type);
        project(configTask);
    });

    // TODO: Set promises right
    promises.push(new Promise(function (resolve) {
        resolve();
    }));

    return Promise.all(promises);
}

// -----------------------------------------
// EXPORTS

module.exports = {
    struct: struct,
    task: task,
    project: project
};
